pySDC/projects/DAE/sweepers/fully_implicit_DAE_MPI.py

```diff
@@ -0,0 +1,229 @@
+        elif L.params.residual_type == 'last_abs':
+            L.status.residual = self.comm.bcast(res_norm, root=self.comm.size - 1)
+        elif L.params.residual_type == 'full_rel':
+            L.status.residual = self.comm.allreduce(res_norm / abs(L.u[0]), op=MPI.MAX)
+        elif L.params.residual_type == 'last_rel':
+            L.status.residual = self.comm.bcast(res_norm / abs(L.u[0]), root=self.comm.size - 1)
+        else:
+            raise NotImplementedError(f'residual type \"{L.params.residual_type}\" not implemented!')
+
+        # indicate that the residual has seen the new values
+        L.status.updated = False
+
+        return None
+
+    def predict(self):
+        r"""
+        Predictor to fill values at nodes before first sweep. It can decide whether the
+
+            - initial condition is spread to each node ('initial_guess' = 'spread'),
+            - or zero values are spread to each node ('initial_guess' = 'zero').
+
+        Default prediction for sweepers, only copies values to all collocation nodes. This function
+        overrides the base implementation by always initialising ``level.f`` to zero. This is necessary since
+        ``level.f`` stores the solution derivative in the fully implicit case, which is not initially known.
+        """
+
+        L = self.level
+        P = L.prob
+
+        # set initial guess for gradient to zero
+        L.f[0] = P.dtype_f(init=P.init, val=0.0)
+
+        if self.params.initial_guess == 'spread':
+            L.u[self.rank + 1] = P.dtype_u(L.u[0])
+            L.f[self.rank + 1] = P.dtype_f(init=P.init, val=0.0)
+        elif self.params.initial_guess == 'zero':
+            L.u[self.rank + 1] = P.dtype_u(init=P.init, val=0.0)
+            L.f[self.rank + 1] = P.dtype_f(init=P.init, val=0.0)
+        else:
+            raise ParameterError(f'initial_guess option {self.params.initial_guess} not implemented')
+
+        # indicate that this level is now ready for sweeps
+        L.status.unlocked = True
+        L.status.updated = True
+
+    def compute_end_point(self):
+        """
+        Compute u at the right point of the interval.
+
+        The value uend computed here is a full evaluation of the Picard formulation unless do_full_update==False
+
+        Returns
+        -------
+        None
+        """
+
+        L = self.level
+        P = L.prob
+        L.uend = P.dtype_u(P.init, val=0.0)
+
+        # check if Mth node is equal to right point and do_coll_update is false, perform a simple copy
+        if self.coll.right_is_node and not self.params.do_coll_update:
+            # a copy is sufficient
+            root = self.comm.Get_size() - 1
+            if self.comm.rank == root:
+                L.uend = P.dtype_u(L.u[-1])
+            # broadcast diff parts and alg parts separately
+            self.comm.Bcast(L.uend.diff, root=root)
+            self.comm.Bcast(L.uend.alg, root=root)
```

Comment:
Again, do you need to do this separately?

Reply:
No, I agree with you and did change that.